public/app/features/explore/hooks/useStateSync.ts

```diff
@@ -0,0 +1,409 @@
+import { isEmpty, isEqual, isObject, mapValues, omitBy } from 'lodash';
+import { useEffect, useRef } from 'react';
+
+import {
+  CoreApp,
+  serializeStateToUrlParam,
+  ExploreUrlState,
+  isDateTime,
+  TimeRange,
+  RawTimeRange,
+  DataSourceApi,
+} from '@grafana/data';
+import { DataQuery, DataSourceRef } from '@grafana/schema';
+import { useGrafana } from 'app/core/context/GrafanaContext';
+import { clearQueryKeys, getLastUsedDatasourceUID, parseUrlState } from 'app/core/utils/explore';
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { MIXED_DATASOURCE_NAME } from 'app/plugins/datasource/mixed/MixedDataSource';
+import { addListener, ExploreId, ExploreItemState, ExploreQueryParams, useDispatch, useSelector } from 'app/types';
+
+import { changeDatasource } from '../state/datasource';
+import { initializeExplore } from '../state/explorePane';
+import { clearPanes, splitClose, splitOpen, syncTimesAction } from '../state/main';
+import { runQueries, setQueriesAction } from '../state/query';
+import { selectPanes } from '../state/selectors';
+import { changeRangeAction, updateTime } from '../state/time';
+import { DEFAULT_RANGE } from '../state/utils';
+import { withUniqueRefIds } from '../utils/queries';
+
+/**
+ * Bi-directionally syncs URL changes with Explore's state.
+ */
+export function useStateSync(params: ExploreQueryParams) {
```

Comment:
Do we need a task for adding unit tests?

Reply:
all the logic in here is already covered by unit tests in ExplorePage.test.tsx. with the migration PR i'll be extracting tests that are URL-specific to their own suite